### PR TITLE
API updates: connect() and more

### PR DESCRIPTION
**Pull request by bpierre**
Documents: https://github.com/aragon/plumbery/tree/docs-connect#documentation

## connect()

The `connect()` function is now the normal way to access an organization, and handles both the connection and loading an organization. The idea here is to optimize for the common scenario of building a thing that only needs to connect to one organization.

It removes the need to have a “library object”, and it makes the connection to the organization more central to the library, especially since the possibility to create an organization has been removed for the time being.

Before:

```js
const plumbery = new Plumbery()
const myorg = plumbery.organization('myorg')
```

After:

```js
const myorg = await connect('myorg')
```

It also has been made asynchronous:

- It gives a chance to connectors to connect (we could have a standard `ConnectorInterface#connect()`).
- It could allow us to retrieve some information about the organization. This is something I was trying to avoid for performance reasons (the connection should be as fast as possible), but it could be kept minimal (e.g. only resolving the ENS name).

Which also means that it can now fail in different ways by throwing errors.

## Errors

I think errors could be put into two buckets: the normal exceptions (e.g. “the org doesn’t seem to exist”), and the ones that shouldn’t happen (e.g. “connect() requires at least one parameter”).

For the ones that shouldn’t happen, I think `throw` / `catch` is an obvious choice.

For other ones, it is common to try to avoid the verbosity of `throw` / `catch` for other solutions. For example, if we take `connect()`, we could return `null` when the organization doesn’t exist:

```js
// Trying to avoid JS errors
import { connect } from 'plumbery'

const myorg = await connect('myorg.aragonid.eth')

if (myorg === null) {
  console.log('The org doesn’t seem to exist!')
}
```

```js
// Using JS errors
import { connect } from 'plumbery'

try {
  const myorg = await connect('myorg.aragonid.eth')
} catch(err) {
  console.log('The org doesn’t seem to exist!')
}
```

Not throwing errors might look a little bit nicer − the first part doesn’t get indented − but it comes with a few issues:

- Things can fail in only one way. For example, being able to make a distinction between a connection error and an organization not being found would require to introduce a new pattern.
- We lose the semantics of errors. A non-intercepted error having for message “The organization doesn’t seem to exist” is more useful than “Can’t access property “name” of `null`”. We don’t strictly need `throw` / `catch` to get the semantics of errors (e.g. Node callbacks or `Promise#then()`), but trying to avoid it would require a non-standard way to pass errors to the caller.
- It doesn’t interrupt the flow like non-catched errors do. Trying to avoid errors with e.g. replacing an object by `null` can create bugs difficult to understand if they cause indirect issues.

For these reasons, I think it could be reasonable to use throw / catch whenever we want to communicate an exception, excluding obvious cases where the returned value makes it really obvious (e.g. an empty array rather than a filled one is generally enough to communicate “no results”).

Now I think the React API could do things a little bit differently, but I will discuss this in another PR.

## IPFS

The connect() function now allows users to define an IPFS gateway resolver.

## Intent

`Intent` is an abstraction that represent anything that will end up in a transaction. Every method resulting in a transaction now returns an intent object.

It gives access to the different transaction paths, but also gives access to the transaction(s) of the shortest one.

## TransactionPath

The transaction path is now represented using a simple data structure. It contains the list of forwarders (`apps`), the last one (`destination`), and the transaction(s) to sign.

We can have multiple transactions in some cases (“pretransaction”) that are not up to the users of the library, so I think we need to communicate this idea of having possibly multiple transactions. It will also allow us to do more things in the future, like being able to assemble a basket of various transactions before signing them all at once.

## TransactionRequest

There is no standard that I know of to represent a transaction request, but Web3.js and ethers.js are both doing very similar things. The name “transaction request” was inspired by the way ethers.js describes it, though they don’t have a type for it.

The structure of this object is a combination of the two formats, and it also has two methods: `#toEthers()` and `#toWeb3()`. Users of these libraries should only have to call these to get the correct format.

### CODE REVIEW

**Review by eternauta1337 — APPROVED** [review]

**Comment by eternauta1337** [comment]
Looking good @bpierre 
Let's merge!

File: docs/connect.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+| ---------------------- | ---------------------------------------------- |
+| `ConnectionError`      | Gets thrown if the connection fails.           |
+| `OrganizationNotFound` | Gets thrown if the organization doesn’t exist. |
+
+### Example
+
+```js
+import { connect } from 'plumbery'
+
+const org = await connect({ connector: 'thegraph' })
```
Comment (kanaagawa) [inline docs/connect.md]: Maybe for the example instead of `org` something like `getOrg`?

File: docs/connect.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  const org2 = await connect('org2.aragonid.eth', { connector: 'thegraph' })
+
+  // Specify a connector and its config
+  const org3 = await connect('org3.aragonid.eth', {
+    connector: ['thegraph', { daoSubgraphUrl: 'http://…' }],
+  })
+
+  // Custom connector
+  const org4 = await connect('org4.aragonid.eth', {
+    connector: new CustomConnector(),
```
Comment (kanaagawa) [inline docs/connect.md:47]: Maybe we could add the`IConnector` the custom connector needs to implement.

File: docs/intent.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Intent
```
Comment (kanaagawa) [inline docs/intent.md:1]: I love how you think about the intent API.

File: docs/intent.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+Intent to change anything on an organization and its apps, through an Ethereum transaction. Allows to get transaction paths from it.
```
Comment (kanaagawa) [inline docs/intent.md]: Wonder if we should mention that intent might also trigger an action on an external contract using Agent.

File: docs/organization.md [hunk 5]
```diff
@@ -87,23 +87,23 @@ Get the current manager of a role.
...
-| `funcArgs`   | `String`                    | Parameters to pass to the function.                   |
-| returns      | `Promise<TransactionPaths>` | A promise resolving to a `TransactionPaths` instance. |
+| Name         | Type              | Description                                  |
+| ------------ | ----------------- | -------------------------------------------- |
+| `appAddress` | `String`          | Address of the app instance.                 |
+| `funcName`   | `String`          | Name of the function to call.                |
+| `funcArgs`   | `String`          | Parameters to pass to the function.          |
+| returns      | `Promise<Intent>` | A promise resolving to an `Intent` instance. |
 
 ### Organization#appCall(appAddress, methodName, args)
```
Comment (kanaagawa) [inline docs/organization.md:108]: I wonder after the discussion on Notion with Brett and today talk with Dan from PieDao maybe we should think about how we want to handle the `appCall` and `appState`. 

On one hand, I think the `appCall` is contrary to what we are trying to do with the graph. Maybe is a good way to have a fallback if some information is not accessible. On the other hand for the Ethereum connector might be duplicating part of the code.

About the `appState` I wonder if we might want to lift this one and handle the `orgState`. I'm afraid if having the `appState` could make cross-app awareness difficult as in the client.

Comment (kanaagawa) [reply docs/organization.md:108]: Another comment is on the `appEvent` I think the concept of an `appEvent` is related in particular with the Ethereum provider. I wonder how that would be represented on the connector for the graph for example.

File: docs/transaction-path.md [hunk 6]
```diff
@@ -2,36 +2,10 @@
 
 Represents a single transaction path.
 
-## Methods
+## Properties
```
Comment (kanaagawa) [inline docs/transaction-path.md:5]: Why you consider them properties?

File: docs/transaction-request.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+| Name       | Type     | Description                                  |
+| ---------- | -------- | -------------------------------------------- |
+| `chainId`  | `Number` | Chain ID of the network.                     |
+| `data`     | `String` | Transaction data.                            |
+| `gasLimit` | `Number` | Maximum gas this transaction may spend.      |
+| `gasPrice` | `String` | Price (in wei) per unit of gas.              |
+| `to`       | `String` | Target address or ENS name.                  |
+| `value`    | `String` | Amount (in wei) this transaction is sending. |
+
+## Methods
```
Comment (kanaagawa) [inline docs/transaction-request.md:16]: Does it make sense to also have `toRpc()` for handling the transaction without a library?

**Review by 0xGabi — APPROVED** [review]
I love the changes. Much more simple and focus on the important aspects. Seems like the `TransactionRequest` is something worth having further discussions on the community and gather consensus on the standard.

File: docs/transaction-path.md (hunk 6 above)
Comment (bpierre) [reply docs/transaction-path.md:5]: I was thinking that we might not want methods anymore on this object, and have everything pre-filled as properties instead. Do you think we should e.g. make `apps` a method to instantiate the `App` objects on demand?

File: docs/transaction-request.md (hunk 7 above)
Comment (bpierre) [reply docs/transaction-request.md:16]: Absolutely, good idea!

File: docs/intent.md (hunk 4 above)
Comment (bpierre) [reply docs/intent.md]: Yes true!

File: docs/connect.md (hunk 2 above)
Comment (bpierre) [reply docs/connect.md:47]: Yes, I try to use JS only in the code examples but we could add it as a comment?

File: docs/transaction-path.md (hunk 6 above)
Comment (kanaagawa) [reply docs/transaction-path.md:5]: I agree, let's keep it simple. 

I just thought, does make sense to have a method to get a gas estimation of the transaction path?

**Comment by 0xGabi** [comment]
I wonder how we plan to represent a transaction path for an intent basket, with the refactor 🤔

**Pull request merged**